kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java

```diff
@@ -0,0 +1,210 @@
+        addSetTransactionIfPresent(metadataActions);
+
+        CloseableIterator<Row> stageDataIter = stagedData.iterator();
+
+        // Create a new CloseableIterator that will return the commit action, metadata action, and
+        // the stagedData
+        CloseableIterator<Row> stagedDataWithCommitInfo = new CloseableIterator<Row>() {
+            private int metadataActionReturned = -1;
+
+            @Override
+            public boolean hasNext() {
+                return metadataActionReturned < metadataActions.size() - 1 ||
+                        stageDataIter.hasNext();
+            }
+
+            @Override
+            public Row next() {
+                if (metadataActionReturned < metadataActions.size() - 1) {
+                    metadataActionReturned += 1;
+                    return metadataActions.get(metadataActionReturned);
+                } else {
+                    return stageDataIter.next();
+                }
+            }
+
+            @Override
+            public void close() throws IOException {
+                stageDataIter.close();
+            }
+        };
+
+        try {
+            long readVersion = readSnapshot.getVersion(tableClient);
+            if (readVersion == -1) {
+                // New table, create a delta log directory
+                tableClient.getFileSystemClient().mkdir(logPath.toString());
+            }
+
+            long newVersion = readVersion + 1;
+            // Write the staged data to a delta file
+            tableClient.getJsonHandler().writeJsonFileAtomically(
+                    FileNames.deltaFile(logPath, newVersion),
+                    stagedDataWithCommitInfo,
+                    false /* overwrite */
+            );
+
+            committed = true;
+
+            return new TransactionCommitStatus(newVersion, readyForCheckpoint(newVersion));
+        } catch (FileAlreadyExistsException e) {
+            // TODO: Resolve conflicts and retry commit
+            throw new ConcurrentWriteException();
+        } catch (IOException ioe) {
+            throw new RuntimeException(ioe);
+        }
+    }
+
+    public Optional<Long> getReadVersion() {
+        return readVersion;
+    }
+
+    public Optional<Predicate> getReadPredicate() {
+        return readPredicate;
+    }
+
+    /**
+     * Get the part of the schema of the table that needs the statistics to be collected per file.
+     *
+     * @param tableClient      {@link TableClient} instance to use.
+     * @param transactionState State of the transaction
+     * @return
+     */
+    public static List<Column> getStatisticsColumns(TableClient tableClient, Row transactionState) {
+        // TODO: implement this once we start supporting collecting stats
+        return Collections.emptyList();
+    }
+
+    private Row generateCommitAction() {
+        return new CommitInfo(System.currentTimeMillis(), engineInfo, operation).toRow();
```

Comment:
Do we insert the kernel-specific engine info? Or do we rely on the connector to do that?